virtool/account/data.py

```diff
@@ -0,0 +1,364 @@
+        :param user_id: the user ID
+        :return: the API key
+        """
+        if isinstance(data.permissions, dict):
+            permissions_dict = data.permissions
+        else:
+            permissions_dict = data.permissions.dict(exclude_unset=True)
+
+        name = data.name
+
+        user = await self._db.users.find_one(
+            user_id, ["administrator", "groups", "permissions"]
+        )
+
+        key_permissions = {
+            **virtool.users.utils.generate_base_permissions(),
+            **permissions_dict,
+        }
+
+        if not user["administrator"]:
+            key_permissions = virtool.users.utils.limit_permissions(
+                key_permissions, user["permissions"]
+            )
+
+        raw, hashed = virtool.utils.generate_key()
+
+        document = {
+            "_id": hashed,
+            "id": await virtool.account.db.get_alternate_id(self._db, name),
+            "name": name,
+            "groups": user["groups"],
+            "permissions": key_permissions,
+            "created_at": virtool.utils.timestamp(),
+            "user": {"id": user_id},
+        }
+
+        await self._db.keys.insert_one(document)
+
+        del document["_id"]
+        del document["user"]
+
+        document["key"] = raw
+
+        return document
+
+    async def delete_keys(self, user_id: str):
+        """
+        Delete all API keys for the account associated with the requesting session.
+
+        :param user_id: the user ID
+        """
+        await self._db.keys.delete_many({"user.id": user_id})
+
+    async def get_key(
+        self, user_id: str, key_id: str, api_key_projection: dict
+    ) -> dict:
+        """
+        Get the complete representation of the API key identified by the `key_id`.
+
+        :param user_id: the user ID
+        :param key_id: the ID of the API key to get
+        :param api_key_projection: the projection for the api key
+        :return: the API key
+        """
+        document = await self._db.keys.find_one(
+            {"id": key_id, "user.id": user_id}, api_key_projection
+        )
+
+        if document is None:
+            raise ResourceNotFoundError()
+
+        return document
+
+    async def edit_key(
+        self,
+        user_id: str,
+        key_id: str,
+        data: EditKeySchema,
+        api_key_projection: dict,
+    ) -> dict:
+        """
+        Change the permissions for an existing API key.
+
+        :param user_id: the user ID
+        :param key_id: the ID of the API key to update
+        :param data: permissions update
+        :param api_key_projection: the projection for the api key
+        :return: the API key
+        """
+        if data.permissions is None:
+            permissions_update = {}
+        else:
+            permissions_update = data.permissions.dict(exclude_unset=True)
+
+        if not await self._db.keys.count_documents({"id": key_id}):
+            raise ResourceNotFoundError()
```

Comment:
See `id_exists` helper function.

Reply:
id_exists wasnt working properly (would return false for an id that DID exist) so I just used count_documents